File: python/blind curated 75/buy_sell_stock.py
def maxProfit(prices):
    left, right = 0, 1
    profit = 0

    while right < len(prices):
        # update left to be the min of the pointers
        if prices[right] < prices[left]:
            left = right
        else:
            # calc profit at all possible right vals
            profit = max(profit, prices[right] - prices[left])

        right += 1

    return profit


print(maxProfit([7, 1, 4, 6, 5]))  # 5
print(maxProfit([7, 6, 5, 4]))  # 0
print(maxProfit([9, 3, 6, 1, 3]))  # 3

File: cpp/blind curated 75/buy_sell_stock.cpp
// Leetcode 121

#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

int maxProfit(vector<int> &prices)
{
    int left = 0, right = 1;
    int profit = 0;

    while (right < prices.size())
    {
        if (prices.at(right) < prices.at(left))
        {
            left = right;
        }
        else
        {
            profit = max(profit, prices.at(right) - prices.at(left));
        }

        right++;
    }

    return profit;
}

int main()
{
    vector<int> v1{7, 1, 4, 6, 5};
    vector<int> v2{7, 6, 5, 4};
    vector<int> v3{9, 3, 6, 1, 3};

    cout << maxProfit(v1) << endl; // 5
    cout << maxProfit(v2) << endl; // 0
    cout << maxProfit(v3) << endl; // 3
}